wasm/jit/engine.go

```diff
@@ -0,0 +1,329 @@
+package jit
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"unsafe"
+
+	"github.com/tetratelabs/wazero/wasm"
+	"github.com/tetratelabs/wazero/wasm/buildoptions"
+)
+
+type engine struct {
+	// The actual Go-allocated stack.
+	// Note that we NEVER edit len or cap in JITed code so we won't get screwed when GC comes in.
+	stack []uint64
+	// Wasm stack pointer on .stack field which is accessed by currentBaseStackPointer+currentStackPointer
+	currentStackPointer     uint64
+	currentBaseStackPointer uint64
+	// Where we store the status code of JIT execution.
+	jitCallStatusCode jitStatusCodes
+	// Set when statusCode == jitStatusCall{Function,BuiltInFunction,HostFunction}
+	// Indicating the function call index.
+	functionCallIndex int64
+	// Set when statusCode == jitStatusCall{Function,BuiltInFunction,HostFunction}
+	// We use this value to continue the current function
+	// after calling the target function exits.
+	// Instructions after [base+continuationAddressOffset] must start with
+	// restoring reserved registeres.
+	continuationAddressOffset uintptr
+	// Function call frames in linked list
+	callFrameStack *callFrame
+	// Store the compiled functions and indexes.
+	compiledWasmFunctions     []*compiledWasmFunction
+	compiledWasmFunctionIndex map[*wasm.FunctionInstance]int64
+	// Store the host functions and indexes.
+	hostFunctions     []func(ctx *wasm.HostFunctionCallContext)
+	hostFunctionIndex map[*wasm.FunctionInstance]int64
+}
+
+func (e *engine) Call(f *wasm.FunctionInstance, args ...uint64) (returns []uint64, err error) {
+	for _, arg := range args {
+		e.push(arg)
+	}
+	if index, ok := e.hostFunctionIndex[f]; ok {
+		e.hostFunctions[index](&wasm.HostFunctionCallContext{Memory: f.ModuleInstance.Memory})
+	} else if index, ok := e.compiledWasmFunctionIndex[f]; ok {
+		f := e.compiledWasmFunctions[index]
+		e.exec(f)
+	} else {
+		err = fmt.Errorf("invalid function")
+		return
+	}
+	returns = make([]uint64, len(f.Signature.ReturnTypes))
```

Comment:
a comment with an example is maybe helpful here as we are using a forward range on returns, but backwards on the stack

Reply:
https://github.com/tetratelabs/wazero/pull/60/commits/a1b00ed9587c44b71bbf04e3d20d18f852b51499 done